base_delivery_carrier_label/models/stock_picking.py

```diff
@@ -0,0 +1,262 @@
+        carrier = self.carrier_id
+        self.carrier_type = carrier.carrier_type
+        self.carrier_code = carrier.code
+        default_options = carrier.default_options()
+        self.option_ids = [(6, 0, default_options.ids)]
+        result = {
+            'domain': {
+                'option_ids': [('id', 'in', carrier.available_option_ids.ids)],
+            }
+        }
+        return result
+
+    @api.onchange('option_ids')
+    def option_ids_change(self):
+        if not self.carrier_id:
+            return
+        carrier = self.carrier_id
+        for available_option in carrier.available_option_ids:
+            if (available_option.mandatory and
+                    available_option not in self.option_ids):
+                # XXX the client does not allow to modify the field that
+                # triggered the onchange:
+                # https://github.com/odoo/odoo/issues/2693#issuecomment-56825399
+                # Ideally we should add the missing option
+                raise UserError(
+                    _("You should not remove a mandatory option."
+                      "Please cancel the edit or "
+                      "add back the option: %s.") % available_option.name
+                )
+
+    @api.model
+    def _values_with_carrier_options(self, values):
+        values = values.copy()
+        carrier_id = values.get('carrier_id')
+        option_ids = values.get('option_ids')
+        if carrier_id and not option_ids:
+            carrier_obj = self.env['delivery.carrier']
+            carrier = carrier_obj.browse(carrier_id)
+            default_options = carrier.default_options()
+            if default_options:
+                values.update(option_ids=[(6, 0, default_options.ids)])
+        return values
+
+    @api.multi
+    @api.returns('stock.quant.package')
+    def _get_packages_from_picking(self):
+        """ Get all the packages from the picking """
+        self.ensure_one()
+        operation_obj = self.env['stock.move.line']
+        packages = self.env['stock.quant.package'].browse()
+        operations = operation_obj.search(
+            ['|',
+             ('package_id', '!=', False),
+             ('result_package_id', '!=', False),
+             ('picking_id', '=', self.id)]
+        )
+        for operation in operations:
+            # Take the destination package. If empty, the package is
+            # moved so take the source one.
+            packages |= operation.result_package_id or operation.package_id
+        return packages
+
+    @api.multi
+    def write(self, vals):
+        """ Set the default options when the delivery method is changed.
+
+        So we are sure that the options are always in line with the
+        current delivery method.
+
+        """
+        vals = self._values_with_carrier_options(vals)
+        return super(StockPicking, self).write(vals)
+
+    @api.model
+    @api.returns('self', lambda value: value.id)
```

Comment:
we return the recordset here

Reply:
Then you suggest to drop this line?
Hasn't it been added for some purpose?

Reply:
IMO is totally useless here

Reply:
OK, removed